### ISSUE TITLE
Cross compiling add some character 

### ISSUE THREAD

**Issue by Ludea** [openssl/openssl#19196]
Hello,
I am building openssl from Windows x86_64 to Android aarch64.
So I have to install cygwin and Perl.
I set `--openssdir` to `/usr/local/ssl`
But get
```
crypto/ct/ct_log.c:143:15: error: expected expression
        fpath = CTLOG_FILE;
                ^
  include/internal/cryptlib.h:61:35: note: expanded from macro 'CTLOG_FILE'
  #  define CTLOG_FILE              OPENSSLDIR "/ct_log_list.cnf"
                                    ^
  <command line>:12:20: note: expanded from here
  #define OPENSSLDIR \\/usr/local/ssl\\\ 
```
The build command add `\\` at start and `\\\` at end, so the openssl dir is wrong.
How can I fix it ?

**Comment by levitte** [comment]
In what shell are you working?  What does the configuration line look like (`perl configdata.pm -c` gives the answer)?

**Comment by Ludea** [comment]
```5m 4s
Run yarn tauri android build
yarn run v1.22.19
$ set TAURI_CC_AARCH64-LINUX-ANDROID=C:/hostedtoolcache/windows/ndk/r25b/x64/toolchains/llvm/prebuilt/windows-x86_64/bin/clang.exe&& node tauri/tooling/cli/node/tauri.js android build
        Info running command "rustc --verbose --version" and waiting for output
        Info detected host target triple "x86_64-pc-windows-msvc"
        Info running command "D:\\a\\tauri-mobile\\tauri-mobile\\src-tauri\\gen\\android\\hello\\gradlew.bat --project-dir D:\\a\\tauri-mobile\\tauri-mobile\\src-tauri\\gen\\android\\hello assembleUniversalRelease -PabiList=arm64-v8a,armeabi-v7a,x86,x86_64 --warn" and waiting for exit
Downloading https://services.gradle.org/distributions/gradle-7.3.3-bin.zip
...........10%...........20%...........30%...........40%...........50%...........60%...........70%...........80%...........90%...........100%
'compileJava' task (current target is 17) and 'compileKotlin' task (current target is 1.8) jvm target compatibility should be set to the same Java version.
Execution optimizations have been disabled for task ':app:rustBuildArm64Release' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: 'D:\a\tauri-mobile\tauri-mobile\src-tauri'. Reason: Task ':app:rustBuildArm64Release' uses this output of task ':app:buildKotlinToolingMetadata' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.
        Info running command "cargo build --package hello --manifest-path D:\\a\\tauri-mobile\\tauri-mobile\\src-tauri\\Cargo.toml --target aarch64-linux-android --release --color always" and waiting for exit
Clang : Clang
target : "aarch64-linux-android"
ndk : 24
    Updating git repository `https://github.com/tauri-apps/tao`
    Updating git repository `https://github.com/tauri-apps/wry`
    Updating git repository `https://github.com/tauri-apps/tauri`
 Downloading crates ...
  Downloaded unicode-ident v1.0.3
  Downloaded bytes v1.2.1
  Downloaded fastrand v1.8.0
  Downloaded itoa v1.0.3
  Downloaded filetime v0.2.17
  Downloaded http-range v0.1.5
  Downloaded ndk v0.6.0
  Downloaded thiserror-impl v1.0.32
  Downloaded thiserror v1.0.32
  Downloaded crypto-common v0.1.6
  Downloaded anyhow v1.0.64
  Downloaded unicode-normalization v0.1.21
  Downloaded ryu v1.0.11
  Downloaded serde_json v1.0.85
  Downloaded proc-macro2 v1.0.43
  Downloaded local-ip-address v0.4.7
  Downloaded raw-window-handle v0.5.0
  Downloaded wildmatch v2.1.1
  Downloaded webview2-com-macros v0.6.0
  Downloaded winres v0.1.12
  Downloaded openssl v0.10.41
  Downloaded num-iter v0.1.43
  Downloaded bytemuck v1.12.1
  Downloaded num_enum_derive v0.5.7
  Downloaded num_enum v0.5.7
  Downloaded slab v0.4.7
  Downloaded smallvec v1.9.0
  Downloaded serde_derive v1.0.144
  Downloaded serde v1.0.144
  Downloaded windows-tokens v0.37.0
  Downloaded quote v1.0.21
  Downloaded openssl-sys v0.9.75
  Downloaded serde_repr v0.1.9
  Downloaded syn v1.0.99
  Downloaded webview2-com v0.19.1
  Downloaded jni-sys v0.3.0
  Downloaded semver v1.0.13
  Downloaded proc-macro-crate v1.2.1
  Downloaded paste v1.0.8
  Downloaded windows-bindgen v0.39.0
  Downloaded serde_urlencoded v0.7.1
  Downloaded combine v4.6.6
  Downloaded android_logger v0.11.1
  Downloaded cargo_toml v0.11.5
  Downloaded jni v0.19.0
  Downloaded alloc-stdlib v0.2.1
  Downloaded alloc-no-stdlib v2.0.3
  Downloaded cpufeatures v0.2.4
  Downloaded crossbeam-utils v0.8.11
  Downloaded crossbeam-channel v0.5.6
  Downloaded brotli-decompressor v2.3.2
  Downloaded libc v0.2.132
  Downloaded tokio v1.20.1
  Downloaded ndk-sys v0.3.0
  Downloaded encoding_rs v0.8.31
  Downloaded windows_x86_64_msvc v0.37.0
  Downloaded cfb v0.6.1
  Downloaded cesu8 v1.1.0
  Downloaded ndk-context v0.1.1
  Downloaded ctor v0.1.23
  Downloaded attohttpc v0.20.0
  Downloaded cty v0.2.2
  Downloaded windows-implement v0.37.0
  Downloaded state v0.5.3
  Downloaded png v0.11.0
  Downloaded ico v0.1.0
  Downloaded inflate v0.3.4
  Downloaded once_cell v1.14.0
  Downloaded generic-array v0.14.6
  Downloaded android_log-sys v0.2.0
  Downloaded deflate v0.7.20
  Downloaded brotli v3.3.4
  Downloaded openssl-src v111.22.0+1.1.1q
  Downloaded windows-metadata v0.39.0
  Downloaded webview2-com-sys v0.19.0
  Downloaded infer v0.7.0
  Downloaded windows v0.37.0
   Compiling proc-macro2 v1.0.43
   Compiling unicode-ident v1.0.3
   Compiling quote v1.0.21
   Compiling syn v1.0.99
   Compiling cfg-if v1.0.0
   Compiling autocfg v1.1.0
   Compiling serde_derive v1.0.144
   Compiling serde v1.0.144
   Compiling ppv-lite86 v0.2.16
   Compiling siphasher v0.3.10
   Compiling libc v0.2.132
   Compiling getrandom v0.1.16
   Compiling memchr v2.5.0
   Compiling proc-macro-hack v0.5.19
   Compiling parking_lot_core v0.9.3
   Compiling once_cell v1.14.0
   Compiling typenum v1.15.0
   Compiling log v0.4.17
   Compiling winapi v0.3.9
   Compiling version_check v0.9.4
   Compiling windows_x86_64_msvc v0.36.1
   Compiling ident_case v1.0.1
   Compiling convert_case v0.4.0
   Compiling fnv v1.0.7
   Compiling strsim v0.10.0
   Compiling smallvec v1.9.0
   Compiling matches v0.1.9
   Compiling scopeguard v1.1.0
   Compiling byteorder v1.4.3
   Compiling new_debug_unreachable v1.0.4
   Compiling itoa v1.0.3
   Compiling precomputed-hash v0.1.1
   Compiling bitflags v1.3.2
   Compiling mac v0.1.1
   Compiling dtoa v0.4.8
   Compiling tinyvec_macros v0.1.0
   Compiling serde_json v1.0.85
   Compiling ryu v1.0.11
   Compiling utf-8 v0.7.6
   Compiling nodrop v0.1.14
   Compiling alloc-no-stdlib v2.0.3
   Compiling stable_deref_trait v1.2.0
   Compiling crc32fast v1.3.2
   Compiling windows_x86_64_msvc v0.39.0
   Compiling bytes v1.2.1
   Compiling itoa v0.4.8
   Compiling cc v1.0.73
   Compiling percent-encoding v2.1.0
   Compiling windows-tokens v0.39.0
   Compiling thin-slice v0.1.1
   Compiling adler32 v1.2.0
   Compiling pkg-config v0.3.25
   Compiling vcpkg v0.2.15
   Compiling unicode-bidi v0.3.8
   Compiling semver v1.0.13
   Compiling anyhow v1.0.64
   Compiling crossbeam-utils v0.8.11
   Compiling futures-core v0.3.23
   Compiling heck v0.4.0
   Compiling jni-sys v0.3.0
   Compiling glob v0.3.0
   Compiling cty v0.2.2
   Compiling regex-syntax v0.6.27
   Compiling futures-task v0.3.23
   Compiling futures-channel v0.3.23
   Compiling pin-project-lite v0.2.9
   Compiling same-file v1.0.6
   Compiling cesu8 v1.1.0
   Compiling lazy_static v1.4.0
   Compiling futures-io v0.3.23
   Compiling adler v1.0.2
   Compiling foreign-types-shared v0.1.1
   Compiling openssl v0.10.41
   Compiling tao v0.13.3 (https://github.com/tauri-apps/tao?branch=dev#816ca49d)
   Compiling uuid v0.8.2
   Compiling futures-util v0.3.23
   Compiling futures-sink v0.3.23
   Compiling wry v0.20.2 (https://github.com/tauri-apps/wry?branch=dev#2bfc6c3d)
   Compiling paste v1.0.8
   Compiling native-tls v0.2.10
   Compiling cpufeatures v0.2.4
   Compiling pin-utils v0.1.0
   Compiling tauri-runtime v0.10.2 (https://github.com/tauri-apps/tauri?branch=next#8ea87e9c)
   Compiling ndk-context v0.1.1
   Compiling fastrand v1.8.0
   Compiling base64 v0.13.0
   Compiling encoding_rs v0.8.31
   Compiling tauri-runtime-wry v0.10.2 (https://github.com/tauri-apps/tauri?branch=next#8ea87e9c)
   Compiling http-range v0.1.5
   Compiling openssl-probe v0.1.5
   Compiling parking v2.0.0
   Compiling wildmatch v2.1.1
   Compiling waker-fn v1.1.0
   Compiling remove_dir_all v0.5.3
   Compiling android_log-sys v0.2.0
   Compiling state v0.5.3
   Compiling getrandom v0.2.7
   Compiling lock_api v0.4.7
   Compiling num-traits v0.2.15
   Compiling num-integer v0.1.45
   Compiling num-iter v0.1.43
   Compiling slab v0.4.7
   Compiling tokio v1.20.1
   Compiling phf_shared v0.8.0
   Compiling phf_shared v0.10.0
   Compiling instant v0.1.12
   Compiling thread_local v1.1.4
   Compiling generic-array v0.14.6
   Compiling fxhash v0.2.1
   Compiling futf v0.1.5
   Compiling tinyvec v1.6.0
   Compiling dtoa-short v0.3.3
   Compiling alloc-stdlib v0.2.1
   Compiling servo_arc v0.1.1
   Compiling openssl-src v111.22.0+1.1.1q
   Compiling form_urlencoded v1.0.1
   Compiling deflate v0.7.20
   Compiling inflate v0.3.4
   Compiling deflate v1.0.0
   Compiling http v0.2.8
   Compiling ndk-sys v0.3.0
   Compiling tauri v1.0.5 (https://github.com/tauri-apps/tauri?branch=next#8ea87e9c)
   Compiling raw-window-handle v0.5.0
   Compiling walkdir v2.3.2
   Compiling miniz_oxide v0.5.3
   Compiling foreign-types v0.3.2
   Compiling cfb v0.6.1
   Compiling rand_core v0.6.3
   Compiling uuid v1.1.2
   Compiling tendril v0.4.3
   Compiling brotli-decompressor v2.3.2
   Compiling unicode-normalization v0.1.21
   Compiling openssl-sys v0.9.75
   Compiling infer v0.7.0
   Compiling rand_chacha v0.3.1
   Compiling rand_core v0.5.1
   Compiling dirs-sys-next v0.1.2
   Compiling xattr v0.2.3
   Compiling num_cpus v1.13.1
   Compiling filetime v0.2.17
   Compiling tempfile v3.3.0
   Compiling aho-corasick v0.7.18
   Compiling combine v4.6.6
   Compiling bstr v0.2.17
   Compiling windows-sys v0.36.1
   Compiling brotli v3.3.4
   Compiling winapi-util v0.1.5
   Compiling idna v0.2.3
   Compiling png v0.17.5
   Compiling flate2 v1.0.24
   Compiling crossbeam-channel v0.5.6
   Compiling futures-lite v1.12.0
   Compiling rand v0.8.5
   Compiling rand_pcg v0.2.1
   Compiling rand_chacha v0.2.2
   Compiling parking_lot v0.12.1
   Compiling dirs-next v2.0.0
   Compiling tar v0.4.38
   Compiling regex v1.6.0
error: failed to run custom build command for `openssl-sys v0.9.75`

Caused by:
  process didn't exit successfully: `D:\a\tauri-mobile\tauri-mobile\src-tauri\target\release\build\openssl-sys-1b6ef9a0ad7a6613\build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=AARCH64_LINUX_ANDROID_OPENSSL_NO_VENDOR
  AARCH64_LINUX_ANDROID_OPENSSL_NO_VENDOR unset
  cargo:rerun-if-env-changed=OPENSSL_NO_VENDOR
  OPENSSL_NO_VENDOR unset
  CC_aarch64-linux-android = None
  CC_aarch64_linux_android = None
  TARGET_CC = Some("C:/hostedtoolcache/windows/ndk/r25b/x64/toolchains/llvm/prebuilt/windows-x86_64/bin/clang.exe")
  CFLAGS_aarch64-linux-android = None
  CFLAGS_aarch64_linux_android = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("false")
  running "perl" "./Configure" "--prefix=D:\\a\\tauri-mobile\\tauri-mobile\\src-tauri\\target\\aarch64-linux-android\\release\\build\\openssl-sys-32f3e8071b4276d9\\out\\openssl-build\\install" "--openssldir=/usr/local/ssl" "no-dso" "no-shared" "no-ssl3" "no-unit-test" "no-comp" "no-zlib" "no-zlib-dynamic" "no-md2" "no-rc5" "no-weak-ssl-ciphers" "no-camellia" "no-idea" "no-seed" "no-stdio" "linux-aarch64" "-O2" "-DANDROID" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=aarch64-linux-android"
  Configuring OpenSSL version 1.1.1q (0x1010111fL) for linux-aarch64
  Using os-specific seed configuration
  Creating configdata.pm
  Creating Makefile
 ```
It run on cygwin terminal.
I run it from GHA https://github.com/Ludea/tauri-mobile/runs/8303989538?check_suite_focus=true#step:15:284

**Comment by Ludea** [comment]
I just see, into make command : `-DOPENSSLDIR="\"/usr/local/ssl\""` whereas i set `--openssdir` to `/usr/local/ssl `. 
`Configure `command add some \

**Comment by Ludea** [comment]
get same issue with powershell

**Comment by levitte** [comment]
> I just see, into make command : `-DOPENSSLDIR="\"/usr/local/ssl\""` whereas i set `--openssdir` to `/usr/local/ssl `. `Configure `command add some \

This is perfectly correct from a Unix shell perspective.  For the compiler, the value *must* be a string, that's what the escaped " (i.e. `\"`) does.  So it seems that the shell used in https://github.com/Ludea/tauri-mobile/runs/8303989538 (PowerShell) uses slightly different semantics for escapes, probably the same as cmd.exe.

It could be that this patch is enough to fix the problem:

``` diff
diff --git a/Configurations/unix-Makefile.tmpl b/Configurations/unix-Makefile.tmpl
index 21fac4ccfb..836038183a 100644
--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -407,10 +407,10 @@ LIB_CPPFLAGS={- our $lib_cppflags =
                           @{$config{lib_cppflags}},
                           @{$config{shared_cppflag}});
                 join(' ', $lib_cppflags,
-                          (map { '-D'.$_ }
-                               'OPENSSLDIR="\"$(OPENSSLDIR)\""',
-                               'ENGINESDIR="\"$(ENGINESDIR)\""',
-                               'MODULESDIR="\"$(MODULESDIR)\""'),
+                          (map { '"-D'.$_.'"' }
+                               'OPENSSLDIR=\"$(OPENSSLDIR)\"',
+                               'ENGINESDIR=\"$(ENGINESDIR)\"',
+                               'MODULESDIR=\"$(MODULESDIR)\"'),
                           '$(CNF_CPPFLAGS)', '$(CPPFLAGS)') -}
 LIB_CFLAGS={- join(' ', $target{lib_cflags} || (),
                         $target{shared_cflag} || (),
```

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of the 3.4 dev cycle barring further updates

**Comment by t8m** [comment]
Duplicate of #24700
